File: dotnet-backend/src/Infrastructure/Services/AiAnalysisService.cs
using Application.DTOs;
using Application.Interfaces;
using Microsoft.Extensions.Configuration;
using System.Net.Http.Json;
using System.Text.Json;
using UglyToad.PdfPig;

namespace Infrastructure.Services;

/// <summary>
/// Service for AI-based document analysis using the Gemini API. Supports both image and text-based files,
/// extracting the relevant content and submitting it to Gemini for classification and data extraction.
/// Fakes a response if the API Key is not configured.
/// </summary>
public class AiAnalysisService(
    IEventLogService eventLogService,
    IConfiguration configuration,
    IHttpClientFactory httpClientFactory) : IAiAnalysisService
{
    /// <summary>
    /// Analyzes a document stream (image or text) with Gemini or mock, returns structured analysis result.
    /// Logs all operations and errors for auditing.
    /// </param>
    /// <param name="fileStream">Input document as stream.</param>
    /// <param name="fileName">Original file name to determine type.</param>
    /// <returns>AnalysisResultDto structed result with either Invoice or Information data filled.</returns>
    public async Task<AnalysisResultDto> AnalyzeDocumentAsync(Stream fileStream, string fileName)
    {
        // Try to get the Gemini API key from configuration
        var apiKey = configuration["Gemini:ApiKey"];
        if (string.IsNullOrEmpty(apiKey))
        {
            // Log warning and use mock analyzer if API key is missing
            await eventLogService.LogEventAsync("AI Analysis Warning", "Gemini API Key not found. Using mock.");
            return await MockAnalyzeAsync(fileName);
        }

        try
        {
            // Determine file type (image or not)
            var extension = Path.GetExtension(fileName).ToLower();
            var isImage = IsImage(extension);

            // List of Gemini message parts for the request (images and/or prompts)
            var parts = new List<object>();

            if (isImage)
            {
                // Reset stream and copy entire image to memory for encoding
                fileStream.Position = 0;
                using var memoryStream = new MemoryStream();
                await fileStream.CopyToAsync(memoryStream);
                var imageBytes = memoryStream.ToArray();
                var base64Image = Convert.ToBase64String(imageBytes);

                // Determine MIME type for Gemini
                var mimeType = extension == ".png" ? "image/png" : "image/jpeg";

                // Add image as inline_data and a user prompt for analysis
                parts.Add(new { inline_data = new { mime_type = mimeType, data = base64Image } });
                parts.Add(new { text = "Analyze this image document." });
            }
            else
            {
                // Try to extract all possible text from the document
                string extractedText = ExtractContent(fileStream, fileName);
                if (string.IsNullOrWhiteSpace(extractedText))
                    throw new Exception("Could not extract text.");

                // Truncate text if very large (Gemini has limits)
                if (extractedText.Length > 30000)
                    extractedText = extractedText.Substring(0, 30000);

                // Add extracted document text for Gemini to analyze
                parts.Add(new { text = $"Analyze this document content:\n\n{extractedText}" });
            }

            // Compose the system prompt with strict JSON instructions for Gemini output structure
            var systemPrompt = @"
You are an expert document analyzer. Analyze the provided document (text or image) and determine if it is an 'Invoice' or 'Information'.
Return ONLY a valid JSON object matching this structure:
{
    ""documentType"": ""Invoice"" | ""Information"",
    ""invoiceData"": {
        ""clientName"": ""string"",
        ""clientAddress"": ""string"",
        ""providerName"": ""string"",
        ""providerAddress"": ""string"",
        ""invoiceNumber"": ""string"",
        ""date"": ""YYYY-MM-DD"",
        ""total"": 0.00,
        ""products"": [
            { ""name"": ""string"", ""quantity"": 0, ""unitPrice"": 0.00, ""total"": 0.00 }
        ]
    },
    ""informationData"": {
        ""description"": ""string"",
        ""summary"": ""string"",
        ""sentiment"": ""Positive"" | ""Negative"" | ""Neutral""
    }
}
If it is an Invoice, populate 'invoiceData' and leave 'informationData' null.
If it is Information, populate 'informationData' and leave 'invoiceData' null.
Ensure dates are valid ISO 8601 strings. Do not use markdown code blocks in response, just raw JSON.
";
            // Insert system prompt at the top of the message parts for Gemini (v1beta doesn't use system prompt field)
            parts.Insert(0, new { text = systemPrompt });

            // Create request body in Gemini v1beta model format
            var requestBody = new
            {
                contents = new[]
                {
                    new { parts = parts.ToArray() }
                },
                generationConfig = new
                {
                    response_mime_type = "application/json" // Explicitly request JSON output
                }
            };

            // Call Gemini API with constructed request and API key
            var client = httpClientFactory.CreateClient();
            var url =
                $"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key={apiKey}";

            var response = await client.PostAsJsonAsync(url, requestBody);

            // Handle HTTP errors with full message for logging
            if (!response.IsSuccessStatusCode)
            {
                var error = await response.Content.ReadAsStringAsync();
                throw new Exception($"Gemini API Error: {response.StatusCode} - {error}");
            }

            // Parse GeminiResponse from returned JSON
            var jsonResponse = await response.Content.ReadFromJsonAsync<GeminiResponse>();
            var textResult = jsonResponse?.Candidates?.FirstOrDefault()?.Content?.Parts?.FirstOrDefault()?.Text;

            if (string.IsNullOrEmpty(textResult))
                throw new Exception("Empty response from Gemini.");

            // Sometimes Gemini ignores prompt and puts JSON in markdown code blocks; clean these
            textResult = textResult.Replace("```json", "").Replace("```", "").Trim();

            // Deserialize the returned JSON string to the expected DTO, with casing tolerant to Gemini
            var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
            var result = JsonSerializer.Deserialize<AnalysisResultDto>(textResult, options);

            if (result == null) throw new Exception("Failed to deserialize AI response.");

            // Prepare summary for log event, with info depending on document type
            var logSummary = result.DocumentType == "Invoice"
                ? $"Invoice {result.InvoiceData?.InvoiceNumber} for {result.InvoiceData?.Total:C}"
                : $"Info: {result.InformationData?.Summary?.Substring(0, Math.Min(50, result.InformationData?.Summary?.Length ?? 0))}...";

            await eventLogService.LogEventAsync("AI Analysis", $"Analyzed {fileName}: {logSummary}");

            return result;
        }
        catch (Exception ex)
        {
            // On failure, log the error and return default "analysis failed" result
            await eventLogService.LogEventAsync("AI Analysis Error", $"Failed to analyze {fileName}: {ex.Message}");
            return new AnalysisResultDto
            {
                DocumentType = "Information",
                InformationData = new InformationData
                {
                    Description = "Analysis Failed",
                    Summary = $"Error: {ex.Message}",
                    Sentiment = "Neutral"
                }
            };
        }
    }

    /// <summary>
    /// Checks if a given file extension corresponds to a supported image format (jpg, jpeg, png).
    /// </summary>
    /// <param name="extension">File extension including leading dot, case insensitive.</param>
    /// <returns>True if supported image extension; false otherwise.</returns>
    private bool IsImage(string extension) => extension == ".jpg" || extension == ".jpeg" || extension == ".png";

    /// <summary>
    /// Extracts readable textual content from the provided stream depending on file type.
    /// Handles PDFs via PdfPig, reads directly for .txt/.csv/.md/.json, or returns file info fallback.
    /// </summary>
    /// <param name="stream">Input file stream (position reset to 0).</param>
    /// <param name="fileName">Filename for extension and fallback.</param>
    /// <returns>Extracted text content or a fallback string if unextractable.</returns>
    private string ExtractContent(Stream stream, string fileName)
    {
        var extension = Path.GetExtension(fileName).ToLower();
        stream.Position = 0;

        if (extension == ".pdf")
        {
            try
            {
                // Extract all text from all pages using PdfPig
                using var document = PdfDocument.Open(stream);
                return string.Join(" ", document.GetPages().Select(p => p.Text));
            }
            catch
            {
                // Return message if unable to extract PDF content
                return "PDF content could not be extracted.";
            }
        }
        else if (extension == ".txt" || extension == ".csv" || extension == ".json" || extension == ".md")
        {
            // Read entire text file into string
            using var reader = new StreamReader(stream, leaveOpen: true);
            return reader.ReadToEnd();
        }

        // Fallback for unsupported types
        return $"[File: {fileName}]";
    }

    /// <summary>
    /// Provides a mock analysis result if Gemini API key is missing or for test purposes.
    /// </summary>
    /// <param name="fileName">The document file name to determine mock logic.</param>
    /// <returns>AnalysisResultDto with either Invoice data or Information data.</returns>
    private async Task<AnalysisResultDto> MockAnalyzeAsync(string fileName)
    {
        await Task.Delay(1500); // Simulate processing delay

        var result = new AnalysisResultDto();

        // File name heuristic: if "invoice" in file name, provide mock invoice data
        if (fileName.ToLower().Contains("invoice"))
        {
            result.DocumentType = "Invoice";
            result.InvoiceData = new InvoiceData
            {
                ClientName = "Tech Solutions Inc.",
                ClientAddress = "123 Innovation Dr",
                ProviderName = "Cloud Services LLC",
                ProviderAddress = "456 Server Ave",
                InvoiceNumber = "INV-MOCK-001",
                Date = DateTime.Now,
                Total = 1500.00m,
                Products = new List<ProductDto>
                {
                    new() { Name = "Mock Service", Quantity = 1, UnitPrice = 1500m, Total = 1500m }
                }
            };
        }
        else
        {
            // Otherwise, provide mock information data
            result.DocumentType = "Information";
            result.InformationData = new InformationData
            {
                Description = "Mock Info",
                Summary = "This is a mock response.",
                Sentiment = "Neutral"
            };
        }

        return result;
    }

    // ==== Gemini API Internal Response Classes ====

    /// <summary>
    /// Represents the root of the Gemini API JSON response
    /// </summary>
    private class GeminiResponse
    {
        public Candidate[]? Candidates { get; set; }
    }

    /// <summary>
    /// Represents a single response candidate returned by Gemini
    /// </summary>
    private class Candidate
    {
        public Content? Content { get; set; }
    }

    /// <summary>
    /// Represents the content payload given by Gemini for a candidate
    /// </summary>
    private class Content
    {
        public Part[]? Parts { get; set; }
    }

    /// <summary>
    /// Represents a part of Gemini's content payload (typically a text result)
    /// </summary>
    private class Part
    {
        public string? Text { get; set; }
    }
}

File: python-backend/src/infrastructure/services/ai_analysis_service.py
import asyncio
import io
import base64
from datetime import datetime
from typing import List, Any
from src.application.interfaces.interfaces import IAiAnalysisService, IEventLogService
from src.application.dtos.ai_analysis_dtos import AnalysisResultDto, InvoiceData, InformationData, ProductDto
from src.config import settings

class AiAnalysisService(IAiAnalysisService):
    """
    Service responsible for analyzing documents (images or text) using a generative AI model (Gemini).
    It evaluates if a document is an Invoice or Information, extracts key fields, and gracefully
    falls back to a local mock if the API is unavailable.
    All analysis attempts are logged via IEventLogService.
    """
    def __init__(self, event_log_service: IEventLogService):
        """
        Initialize the AI analysis service.

        Args:
            event_log_service (IEventLogService): Service used for logging analytics and errors.
        """
        self.event_log_service = event_log_service

    async def analyze_document(self, file_content: bytes, file_name: str) -> Any:
        """
        Analyze a document (image or text-based) for semantic classification (Invoice/Information) and extract data.
        Relies on Gemini API if available, otherwise uses a mock method.

        Args:
            file_content (bytes): The raw content of the document file.
            file_name (str): The name of the uploaded file.

        Returns:
            AnalysisResultDto: Structured result of analysis (Invoice fields or Information summary).
        """
        api_key = settings.GEMINI_API_KEY
        
        # Use fallback if API key is missing
        if not api_key:
            await self.event_log_service.log_event("AI Analysis Warning", "Gemini API Key not found. Using mock.")
            return await self._mock_analyze(file_name)

        try:
            import google.generativeai as genai
            import json
            
            genai.configure(api_key=api_key)
            model = genai.GenerativeModel('gemini-2.5-flash')
            
            # Determine file type (image vs. text)
            extension = file_name.lower().split('.')[-1] if '.' in file_name else ""
            is_image = self._is_image(extension)
            
            parts = []
            
            # ----- SYSTEM PROMPT DESIGN -----
            # Provides clear JSON contract to the AI model
            system_prompt = """
You are an expert document analyzer. Analyze the provided document (text or image) and determine if it is an 'Invoice' or 'Information'.
Return ONLY a valid JSON object matching this structure:
{
    "documentType": "Invoice" | "Information",
    "invoiceData": {
        "clientName": "string",
        "clientAddress": "string",
        "providerName": "string",
        "providerAddress": "string",
        "invoiceNumber": "string",
        "date": "YYYY-MM-DD",
        "total": 0.00,
        "products": [
            { "name": "string", "quantity": 0, "unitPrice": 0.00, "total": 0.00 }
        ]
    },
    "informationData": {
        "description": "string",
        "summary": "string",
        "sentiment": "Positive" | "Negative" | "Neutral"
    }
}
If it is an Invoice, populate 'invoiceData' and leave 'informationData' null.
If it is Information, populate 'informationData' and leave 'invoiceData' null.
Ensure dates are valid ISO 8601 strings. Do not use markdown code blocks in response, just raw JSON.
"""
            parts.append(system_prompt)

            if is_image:
                # Build image payload for Gemini API (set appropriate image MIME type)
                mime_type = "image/png" if extension == "png" else "image/jpeg"
                parts.append({
                    "mime_type": mime_type,
                    "data": file_content
                })
                parts.append("Analyze this image document.")
            else:
                # Extract text content from file; raise if not possible
                extracted_text = self._extract_content(file_content, file_name)
                if not extracted_text or not extracted_text.strip():
                    raise Exception("Could not extract text.")
                
                # Limit text length to avoid API overload
                if len(extracted_text) > 30000:
                    extracted_text = extracted_text[:30000]
                
                parts.append(f"Analyze this document content:\n\n{extracted_text}")
            
            # ----- GEMINI API CALL -----
            response = model.generate_content(parts)
            response_text = response.text
            
            # ----- Postprocess AI Output for JSON Only -----
            # Remove any accidental markdown or code fences around the JSON that Gemini may return
            if "```json" in response_text:
                response_text = response_text.split("```json")[1].split("```")[0]
            elif "```" in response_text:
                response_text = response_text.split("```")[1].split("```")[0]
                
            # Parse JSON result
            data = json.loads(response_text.strip())
            
            # Convert dict to DTO (will validate fields)
            result = AnalysisResultDto(**data)
            
            # ----- LOGGING -----
            log_summary = ""
            if result.document_type == "Invoice" and result.invoice_data:
                log_summary = f"Invoice {result.invoice_data.invoice_number} for {result.invoice_data.total}"
            elif result.information_data:
                summary_text = result.information_data.summary or ""
                log_summary = f"Info: {summary_text[:50]}..."
                
            await self.event_log_service.log_event("AI Analysis", f"Analyzed {file_name}: {log_summary}")
            
            return result
            
        except Exception as e:
            # Handle API/model errors gracefully and log failure event
            await self.event_log_service.log_event("AI Analysis Error", f"Failed to analyze {file_name}: {str(e)}")
            print(f"Gemini Error: {e}")
            # Return fallback "analysis failed" info type result for visibility
            return AnalysisResultDto(
                document_type="Information",
                information_data=InformationData(
                    description="Analysis Failed",
                    summary=f"Error: {str(e)}",
                    sentiment="Neutral"
                )
            )

    def _is_image(self, extension: str) -> bool:
        """
        Helper to check if the file extension is a supported image type.

        Args:
            extension (str): The file extension.

        Returns:
            bool: True if the extension is an image, else False.
        """
        return extension in ["jpg", "jpeg", "png"]

    def _extract_content(self, file_content: bytes, file_name: str) -> str:
        """
        Extracts text content from a file based on its format.

        - For PDFs, uses pypdf to extract all pages.
        - For plain text formats (txt/csv/json/md), decodes as UTF-8.
        - Returns filename string for unrecognized types.

        Args:
            file_content (bytes): Raw file bytes
            file_name (str): Name of uploaded file

        Returns:
            str: Extracted text content or error placeholder.
        """
        extension = file_name.lower().split('.')[-1] if '.' in file_name else ""
        
        if extension == "pdf":
            try:
                import pypdf
                pdf_file = io.BytesIO(file_content)
                reader = pypdf.PdfReader(pdf_file)
                text = ""
                for page in reader.pages:
                    text += page.extract_text() + " "
                return text
            except Exception as e:
                print(f"PDF Extraction Error: {e}")
                return "PDF content could not be extracted."
        
        elif extension in ["txt", "csv", "json", "md"]:
            try:
                return file_content.decode('utf-8', errors='ignore')
            except:
                return ""
        
        # If file could not be parsed, return its name for display/logging
        return f"[File: {file_name}]"

    async def _mock_analyze(self, file_name: str) -> AnalysisResultDto:
        """
        Fake/mock analysis for local development and when the AI API key is missing.
        Synthesizes an Invoice or Information result based on file_name.

        Args:
            file_name (str): The uploaded file name.

        Returns:
            AnalysisResultDto: Mock analysis result matching the DTO contract.
        """
        await asyncio.sleep(1.5)  # Simulate async delay
        if "invoice" in file_name.lower():
            return AnalysisResultDto(
                document_type="Invoice",
                invoice_data=InvoiceData(
                    client_name="Tech Solutions Inc.",
                    client_address="123 Innovation Dr",
                    provider_name="Cloud Services LLC",
                    provider_address="456 Server Ave",
                    invoice_number="INV-MOCK-001",
                    date=datetime.now(),
                    total=1500.00,
                    products=[
                        ProductDto(name="Mock Service", quantity=1, unit_price=1500.0, total=1500.0)
                    ]
                )
            )
        else:
            return AnalysisResultDto(
                document_type="Information",
                information_data=InformationData(
                    description="Mock Info",
                    summary="This is a mock response.",
                    sentiment="Neutral"
                )
            )
